hipcheck/src/metric/entropy.rs

```diff
@@ -150,8 +152,78 @@ fn is_likely_source_file(
 
 /// Calculate grapheme frequencies for each commit.
 fn grapheme_freqs(commit_diff: &CommitDiff, db: &dyn MetricProvider) -> Result<CommitGraphemeFreq> {
-	let grapheme_freqs = GraphemeFreqCalculator::for_diff(&commit_diff.diff, db)?.calculate();
+	// Dashmap (fast concurrent hashmap) to store counts for each grapheme.
+	let grapheme_table: DashMap<String, u64> = DashMap::new();
 
+	// Iterate over the file diffs by reference.
+	let mut res: Vec<usize> = vec![];
+	let tgt_diffs: Result<Vec<&FileDiff>> = commit_diff
+		.diff
+		.file_diffs
+		.iter()
+		.filter_map(|file_diff| {
+			// Filter out any that are probably not source files, or are empty patches
+			let is_source: bool = match db.is_likely_source_file(Arc::clone(&file_diff.file_name)) {
+				Err(e) => {
+					return Some(Err(e));
+				}
+				Ok(s) => s,
+			};
+			if !is_source || file_diff.patch.is_empty() {
+				None
+			} else {
+				Some(Ok(file_diff))
+			}
+		})
+		.collect();
+	tgt_diffs?
+		.par_iter()
+		.map(|file_diff| {
+			// Count the number of graphemes in this patch, add it to the total,
+			// and track the number of each grapheme.
+			file_diff
+				.patch
+				// Iterate in parallel over the lines of the patch.
+				.par_lines()
```

Comment:
It's probably preferable to `flat_map()` through to `par_lines()` here, if we're doing everything in parallel.